### PR TITLE
feat: add lint `struct_marked_non_exhaustive_changed_type`

### PR DESCRIPTION
**Pull request by CommanderStorm**
This implemnents the second part of https://github.com/obi1kenobi/cargo-semver-checks/issues/954

Resolves https://github.com/obi1kenobi/cargo-semver-checks/issues/954

TODO:
- [x] The implementation is based on https://github.com/obi1kenobi/cargo-semver-checks/pull/962, to remove the merge conflict once that is merged ^^
      => will need to rebase before this is reviewable
- [x] Self review tomorrow

### CODE REVIEW

File: src/lints/struct_marked_non_exhaustive_changed_type.ron [hunk 1]
```diff
@@ -0,0 +1,74 @@
+SemverQuery(
+    id: "struct_marked_non_exhaustive_changed_type",
+    human_readable_name: "struct marked #[non_exhaustive] with no fields of any kind became an enum or union",
+    description: "A struct marked #[non_exhaustive] with no fields of any kind was converted into an enum or union, breaking pattern matching.",
```
Comment (obi1kenobi) [inline src/lints/struct_marked_non_exhaustive_changed_type.ron:4]: nitpick: "of any kind" will probably be confusing to our users. Enum variants have kinds: plain, tuple, or struct variant. Struct fields don't really have a "kind" — they are merely public or not.

Let's avoid mentions of "kind" throughout.
```suggestion
    human_readable_name: "struct marked #[non_exhaustive] with no fields became an enum or union",
    description: "A struct marked #[non_exhaustive] with no fields was converted into an enum or union, breaking pattern matching.",
```

File: src/lints/struct_marked_non_exhaustive_changed_type.ron [hunk 2]
```diff
@@ -0,0 +1,74 @@
+SemverQuery(
+    id: "struct_marked_non_exhaustive_changed_type",
```
Comment (obi1kenobi) [inline src/lints/struct_marked_non_exhaustive_changed_type.ron:2]: Unfortunately this name is misleading. The other `marked_non_exhaustive` lints are about *adding* `#[non_exhaustive]` to a type. The "marked" is the action that caused the breakage.

This name seems to follow the pattern until at the last second it adds "changed type" which will be confusing to our users.

How about naming this lint `non_exhaustive_struct_changed_type`? That follows the precedent of `constructible_struct_changed_type` and other "<optional adjective> <subject> <action causing breakage>" lints?

File: src/lints/struct_marked_non_exhaustive_changed_type.ron [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+            }
+        }
+    }"#,
+    arguments: {
+        "public": "public",
+        "non_exhaustive": "#[non_exhaustive]",
+        "true": true,
+        "zero": 0,
+    },
+    error_message: "A struct marked #[non_exhaustive] with no fields of any kind was converted into an enum or union, breaking pattern matching",
```
Comment (obi1kenobi) [inline src/lints/struct_marked_non_exhaustive_changed_type.ron:69]:
```suggestion
    error_message: "A struct marked #[non_exhaustive] with no fields was converted into an enum or union, breaking pattern matching",
```

**Review by obi1kenobi — COMMENTED** [review]
Nice work! Just a couple of minor nitpicks and this is good to merge.

File: test_outputs/query_execution/struct_marked_non_exhaustive_changed_type.snap [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+      "struct_type": String("plain"),
+      "struct_typename": String("Struct"),
+      "visibility_limit": String("public"),
+    },
+    {
+      "current_typename": String("Enum"),
+      "name": String("NoneExhaustiveFieldlessUnit"),
+      "path": List([
+        String("struct_becomes_enum"),
+        String("NoneExhaustiveFieldlessUnit"),
```
Comment (CommanderStorm) [inline test_outputs/query_execution/struct_marked_non_exhaustive_changed_type.snap:25]:
```suggestion
      "name": String("NonExhaustiveFieldlessUnit"),
      "path": List([
        String("struct_becomes_enum"),
        String("NonExhaustiveFieldlessUnit"),
```

File: test_outputs/witnesses/struct_marked_non_exhaustive_changed_type.snap [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+---
+[["./test_crates/struct_becomes_enum/"]]
+filename = 'src/lib.rs'
+begin_line = 82
+hint = 'matches!(value, struct_becomes_enum::NonExhaustiveEmptyStructToEnum {..});'
+
+[["./test_crates/struct_becomes_enum/"]]
+filename = 'src/lib.rs'
+begin_line = 95
+hint = 'matches!(value, struct_becomes_enum::NoneExhaustiveFieldlessUnit {..});'
```
Comment (CommanderStorm) [inline test_outputs/witnesses/struct_marked_non_exhaustive_changed_type.snap:14]:
```suggestion
hint = 'matches!(value, struct_becomes_enum::NonExhaustiveFieldlessUnit {..});'
```

File: src/query.rs [hunk 6]
```diff
@@ -1009,7 +1009,7 @@ add_lints!(
     constructible_struct_adds_field,
     constructible_struct_adds_private_field,
     constructible_struct_changed_type,
-    struct_marked_non_exhaustive_changed_type,
+    non_exhaustive_struct_changed_type,
```
Comment (obi1kenobi) [inline src/query.rs:1012]: Please keep these sorted, or else our automation will break.

Comment (obi1kenobi) [reply src/query.rs:1012]: See the comment above the macro

Comment (CommanderStorm) [reply src/query.rs]: Good catch

**Comment by CommanderStorm** [comment]
> think this still needs a snapshot file update to pass CI

Oops, missed those files / forgot to change the testcases during the renaming

**Pull request merged**